src/lib/utilities/sokoban.ts

```diff
@@ -0,0 +1,512 @@
+			// Set the player to the next position
+			this.set(
+				nextPosition.x,
+				nextPosition.y,
+				nextPosition.component === EmojiGameComponent.BoxTarget ? EmojiGameComponent.PlayerTarget : EmojiGameComponent.Player
+			);
+
+			// Get the next position of the box
+			const nextBoxPosition = {
+				x: nextPosition.x + (direction === Direction.Left ? -1 : direction === Direction.Right ? 1 : 0),
+				y: nextPosition.y + (direction === Direction.Up ? -1 : direction === Direction.Down ? 1 : 0)
+			};
+			const nextBoxPositionComponent = this.get(nextBoxPosition.x, nextBoxPosition.y);
+
+			// Set the box to the next position
+			this.set(
+				nextBoxPosition.x,
+				nextBoxPosition.y,
+				nextBoxPositionComponent.component === EmojiGameComponent.FloorTarget ? EmojiGameComponent.BoxTarget : EmojiGameComponent.Box
+			);
+			return Option.some(this.get(nextBoxPosition.x, nextBoxPosition.y));
+		}
+
+		// Check if the next position is a floor target, if so set the player as a player target
+		if (nextPosition.component === EmojiGameComponent.FloorTarget) {
+			this.set(nextPosition.x, nextPosition.y, EmojiGameComponent.PlayerTarget);
+			return Option.none;
+		}
+
+		// Set the player to the next position
+		this.set(nextPosition.x, nextPosition.y, EmojiGameComponent.Player);
+		return Option.none;
+	}
+
+	/**
+	 * Determines whether the win condition has been met.
+	 *
+	 * Cases that nullify the win condition:
+	 * - A box isn't on a target
+	 * - A floor target remains
+	 * - The player is on a target
+	 */
+	public checkWinCondition() {
+		return !this.gameComponents.some((c) =>
+			[EmojiGameComponent.FloorTarget, EmojiGameComponent.Box, EmojiGameComponent.PlayerTarget].includes(c)
+		);
+	}
+
+	/**
+	 * Determines whether the box has been cornered.
+	 *
+	 * @param pushedBox the box that was pushed.
+	 *
+	 * Possible cases for the box being cornered are:
+	 * - top and left
+	 * - top and right
+	 * - bottom and left
+	 * - bottom and right
+	 *
+	 * # Note:
+	 * The box can get stuck between boxes too but that gets a lot more complicated to check
+	 * whether the boxes/walls it's stuck between are in an unreachable spot (i.e. accidentally blocking off a path)
+	 */
+	public checkLoseCondition(pushedBox: GameComponent) {
+		if (pushedBox.component !== EmojiGameComponent.Box) return false;
+
+		const [top, bottom, left, right] = [
+			this.get(pushedBox.x, pushedBox.y - 1),
+			this.get(pushedBox.x, pushedBox.y + 1),
+			this.get(pushedBox.x - 1, pushedBox.y),
+			this.get(pushedBox.x + 1, pushedBox.y)
+		];
+
+		return (
+			(top?.component === EmojiGameComponent.Wall && left?.component === EmojiGameComponent.Wall) ||
+			(top?.component === EmojiGameComponent.Wall && right?.component === EmojiGameComponent.Wall) ||
+			(bottom?.component === EmojiGameComponent.Wall && left?.component === EmojiGameComponent.Wall) ||
+			(bottom?.component === EmojiGameComponent.Wall && right?.component === EmojiGameComponent.Wall)
```

Comment:
None of the variables are nullable, only `x.component` is, which will never happen in this case.